onnxruntime/core/providers/cuda/cupti_manager.cc

```diff
@@ -0,0 +1,163 @@
+    CUptiResult status;
+    do {
+      EventRecord event;
+      status = cuptiActivityGetNextRecord(reinterpret_cast<uint8_t*>(const_cast<char*>(buffer.GetData())), size, &record);
+      if (status == CUPTI_SUCCESS) {
+        if (CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL == record->kind ||
+            CUPTI_ACTIVITY_KIND_KERNEL == record->kind) {
+          CUpti_ActivityKernel3* kernel = (CUpti_ActivityKernel3*)record;
+          std::unordered_map<std::string, std::string> args{
+              {"stream", std::to_string(kernel->streamId)},
+              {"grid_x", std::to_string(kernel->gridX)},
+              {"grid_y", std::to_string(kernel->gridY)},
+              {"grid_z", std::to_string(kernel->gridZ)},
+              {"block_x", std::to_string(kernel->blockX)},
+              {"block_y", std::to_string(kernel->blockY)},
+              {"block_z", std::to_string(kernel->blockZ)},
+          };
+
+          std::string name{demangle(kernel->name)};
+
+          new (&event) EventRecord{
+              /* cat = */ EventCategory::KERNEL_EVENT,
+              /* pid = */ -1,
+              /* tid = */ -1,
+              /* name = */ std::move(name),
+              /* ts = */ (int64_t)(kernel->start - start_time_ns) / 1000,
+              /* dur = */ (int64_t)(kernel->end - kernel->start) / 1000,
+              /* args = */ std::move(args)};
+          MapEventToClient(kernel->correlationId, std::move(event));
+        } else if (CUPTI_ACTIVITY_KIND_MEMCPY == record->kind) {
+          CUpti_ActivityMemcpy* mmcpy = (CUpti_ActivityMemcpy*)record;
+          std::string name{GetMemcpyKindString((CUpti_ActivityMemcpyKind)mmcpy->copyKind)};
+          std::unordered_map<std::string, std::string> args{
+              {"stream", std::to_string(mmcpy->streamId)},
+              {"grid_x", "-1"},
+              {"grid_y", "-1"},
+              {"grid_z", "-1"},
+              {"block_x", "-1"},
+              {"block_y", "-1"},
+              {"block_z", "-1"},
+          };
+          new (&event) EventRecord{
+              /* cat = */ EventCategory::KERNEL_EVENT,
+              /* pid = */ -1,
+              /* tid = */ -1,
+              /* name = */ std::move(name),
+              /* ts = */ (int64_t)(mmcpy->start - start_time_ns) / 1000,
+              /* dur = */ (int64_t)(mmcpy->end - mmcpy->start) / 1000,
+              /* args = */ std::move(args)};
+          MapEventToClient(mmcpy->correlationId, std::move(event));
+        } else if (CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION == record->kind) {
+          auto correlation = reinterpret_cast<const CUpti_ActivityExternalCorrelation*>(record);
+          NotifyNewCorrelation(correlation->correlationId, correlation->externalId);
+        }
+      }
+    } while (status == CUPTI_SUCCESS);
+  } /* for */
+}
+
+void CUPTIAPI CUPTIManager::BufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords) {
+  uint8_t* bfr = (uint8_t*)malloc(kActivityBufferSize + kActivityBufferAlignSize);
```

Comment:
Isn't it a memory leak? There is no way to call "free(bfr)"

Reply:
Indeed it is. Thank you for catching it. The alignment itself is unnecessary as I'd noted in the header. I've just deleted this alignment bit.